### PR TITLE
Argument parser

### PR DESCRIPTION
**Pull request by camilbancioiu**

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+// ErrNilArguments signals that arguments from transactions data is nil
+var ErrNilArguments = errors.New("smart contract arguments are nil")
+
+// ErrNilCode signals that code from transaction data is nil
+var ErrNilCode = errors.New("smart contract code is nil")
+
+// ErrNilFunction signals that function from transaction data is nil
+var ErrNilFunction = errors.New("smart contract function is nil")
+
```
Comment (sasurobert) [inline errors.go]: comment is missing.

File: atArgumentParser.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		if i < len(storageUpdates)-1 {
+			data = data + at.GetSeparator()
+		}
+	}
+	return data
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (at *atArgumentParser) IsInterfaceNil() bool {
+	if at == nil {
```
Comment (sasurobert) [inline atArgumentParser.go]: return at == nil

File: errors.go (hunk 1 above)
Comment (camilbancioiu) [reply errors.go]: Added.

File: atArgumentParser.go (hunk 2 above)
Comment (camilbancioiu) [reply atArgumentParser.go]: Changed.

File: atArgumentParser.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
+package vmcommon
+
+import (
+	"encoding/hex"
+	"strings"
+)
+
+// ArgumentsParser defines the functionality to parse transaction data into arguments and code for smart contracts
+type ArgumentsParser interface {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply atArgumentParser.go]: Removed, indeed.

File: atArgumentParser.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+type atArgumentParser struct {
+	arguments [][]byte
+	code      []byte
+}
+
+const atSep = "@"
+const atSepChar = '@'
+
+// NewAtArgumentParser creates a new argument parser implementation that splits arguments by @ character
+func NewAtArgumentParser() (ArgumentsParser, error) {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply atArgumentParser.go]: Fixed.

**Pull request merged**